File: c/src/safety_child.c
/*
 * safety_child.c — CircleAI.Safety.Child domain primitives (C11 port).
 *
 * Ports ChildSafetyPrimitives.cs: TrustedAdult / Geofence / CheckIn records
 * and InMemoryChildSafetyBoard. Adults + fences are keyed dictionaries
 * (last-write-wins); check-ins are an append list. RingOrdered sorts by
 * RingPriority ascending (stable); RecentCheckIns filters by ChildId and
 * returns the newest `limit` descending; IsInsideAnyFence uses the same
 * Haversine formula (R = 6_371_000 m) as the C#.
 *
 * Pure C11 + libc + libm.
 */

#include "circle_ai/safety_child.h"

#include <stdlib.h>
#include <string.h>
#include <math.h>

/* Local PI (matches the C# Math.PI double); avoids relying on the non-standard
 * M_PI from <math.h>, which is unavailable under strict -std=c11. */
#define SC_PI 3.14159265358979323846

static char *sc_strdup(const char *s) {
    if (!s) return NULL;
    size_t n = strlen(s) + 1;
    char *p = (char *)malloc(n);
    if (p) memcpy(p, s, n);
    return p;
}

/* ── records ────────────────────────────────────────────────────────────── */

void ca_trusted_adult_free(ca_trusted_adult_t *a) {
    if (!a) return;
    free(a->adult_id);
    free(a->name);
    free(a->phone);
    free(a->relationship);
    a->adult_id = a->name = a->phone = a->relationship = NULL;
}
void ca_trusted_adult_free_array(ca_trusted_adult_t *arr, size_t count) {
    if (!arr) return;
    for (size_t i = 0; i < count; ++i) ca_trusted_adult_free(&arr[i]);
    free(arr);
}
ca_trusted_adult_t *ca_trusted_adult_copy(ca_trusted_adult_t *dst,
                                          const ca_trusted_adult_t *src) {
    if (!dst || !src) return dst;
    dst->adult_id      = sc_strdup(src->adult_id);
    dst->name          = sc_strdup(src->name);
    dst->phone         = sc_strdup(src->phone);
    dst->relationship  = sc_strdup(src->relationship);
    dst->ring_priority = src->ring_priority;
    return dst;
}

void ca_geofence_free(ca_geofence_t *g) {
    if (!g) return;
    free(g->fence_id);
    free(g->name);
    g->fence_id = g->name = NULL;
}
ca_geofence_t *ca_geofence_copy(ca_geofence_t *dst, const ca_geofence_t *src) {
    if (!dst || !src) return dst;
    dst->fence_id      = sc_strdup(src->fence_id);
    dst->name          = sc_strdup(src->name);
    dst->centre_lat    = src->centre_lat;
    dst->centre_lon    = src->centre_lon;
    dst->radius_meters = src->radius_meters;
    return dst;
}

void ca_check_in_free(ca_check_in_t *c) {
    if (!c) return;
    free(c->child_id);
    free(c->status);
    c->child_id = c->status = NULL;
}
void ca_check_in_free_array(ca_check_in_t *arr, size_t count) {
    if (!arr) return;
    for (size_t i = 0; i < count; ++i) ca_check_in_free(&arr[i]);
    free(arr);
}
ca_check_in_t *ca_check_in_copy(ca_check_in_t *dst, const ca_check_in_t *src) {
    if (!dst || !src) return dst;
    dst->child_id  = sc_strdup(src->child_id);
    dst->status    = sc_strdup(src->status);
    dst->has_lat   = src->has_lat;
    dst->lat       = src->lat;
    dst->has_lon   = src->has_lon;
    dst->lon       = src->lon;
    dst->at_utc_ms = src->at_utc_ms;
    return dst;
}

/* ── board ──────────────────────────────────────────────────────────────── */

struct ca_child_safety_board {
    ca_trusted_adult_t *adults;    /* keyed by adult_id */
    size_t              adu_count, adu_cap;
    ca_geofence_t      *fences;    /* keyed by fence_id */
    size_t              fen_count, fen_cap;
    ca_check_in_t      *checkins;  /* append list */
    size_t              chk_count, chk_cap;
};

ca_child_safety_board_t *ca_child_safety_board_create(void) {
    return (ca_child_safety_board_t *)calloc(1, sizeof(ca_child_safety_board_t));
}
void ca_child_safety_board_destroy(ca_child_safety_board_t *board) {
    if (!board) return;
    for (size_t i = 0; i < board->adu_count; ++i) ca_trusted_adult_free(&board->adults[i]);
    free(board->adults);
    for (size_t i = 0; i < board->fen_count; ++i) ca_geofence_free(&board->fences[i]);
    free(board->fences);
    for (size_t i = 0; i < board->chk_count; ++i) ca_check_in_free(&board->checkins[i]);
    free(board->checkins);
    free(board);
}

bool ca_child_safety_board_add_adult(ca_child_safety_board_t *board,
                                     const ca_trusted_adult_t *a) {
    if (!board || !a) return false;
    for (size_t i = 0; i < board->adu_count; ++i) {
        if (board->adults[i].adult_id && a->adult_id &&
            strcmp(board->adults[i].adult_id, a->adult_id) == 0) {
            ca_trusted_adult_t copy; memset(&copy, 0, sizeof(copy));
            ca_trusted_adult_copy(&copy, a);
            ca_trusted_adult_free(&board->adults[i]);
            board->adults[i] = copy;
            return true;
        }
    }
    if (board->adu_count == board->adu_cap) {
        size_t nc = board->adu_cap ? board->adu_cap * 2 : 8;
        void *n = realloc(board->adults, nc * sizeof(*board->adults));
        if (!n) return false;
        board->adults = n; board->adu_cap = nc;
    }
    ca_trusted_adult_copy(&board->adults[board->adu_count], a);
    board->adu_count++;
    return true;
}

ca_trusted_adult_t *ca_child_safety_board_ring_ordered(ca_child_safety_board_t *board,
                                                       size_t *out_count) {
    if (out_count) *out_count = 0;
    if (!board) { if (out_count) *out_count = SIZE_MAX; return NULL; }
    size_t n = board->adu_count;
    if (n == 0) return NULL;
    /* OrderBy(a => a.RingPriority) — ascending, stable for ties. */
    size_t *idx = (size_t *)malloc(n * sizeof(size_t));
    if (!idx) { if (out_count) *out_count = SIZE_MAX; return NULL; }
    for (size_t i = 0; i < n; ++i) idx[i] = i;
    for (size_t i = 1; i < n; ++i) {
        size_t cur = idx[i];
        int key = board->adults[cur].ring_priority;
        size_t j = i;
        while (j > 0 && board->adults[idx[j - 1]].ring_priority > key) { idx[j] = idx[j - 1]; --j; }
        idx[j] = cur;
    }
    ca_trusted_adult_t *res = (ca_trusted_adult_t *)calloc(n, sizeof(*res));
    if (!res) { free(idx); if (out_count) *out_count = SIZE_MAX; return NULL; }
    for (size_t i = 0; i < n; ++i) ca_trusted_adult_copy(&res[i], &board->adults[idx[i]]);
    free(idx);
    if (out_count) *out_count = n;
    return res;
}

bool ca_child_safety_board_define_geofence(ca_child_safety_board_t *board,
                                           const ca_geofence_t *g) {
    if (!board || !g) return false;
    for (size_t i = 0; i < board->fen_count; ++i) {
        if (board->fences[i].fence_id && g->fence_id &&
            strcmp(board->fences[i].fence_id, g->fence_id) == 0) {
            ca_geofence_t copy; memset(&copy, 0, sizeof(copy));
            ca_geofence_copy(&copy, g);
            ca_geofence_free(&board->fences[i]);
            board->fences[i] = copy;
            return true;
        }
    }
    if (board->fen_count == board->fen_cap) {
        size_t nc = board->fen_cap ? board->fen_cap * 2 : 8;
        void *n = realloc(board->fences, nc * sizeof(*board->fences));
        if (!n) return false;
        board->fences = n; board->fen_cap = nc;
    }
    ca_geofence_copy(&board->fences[board->fen_count], g);
    board->fen_count++;
    return true;
}

bool ca_child_safety_board_get_geofence(ca_child_safety_board_t *board,
                                        const char *id, ca_geofence_t *out) {
    if (!board || !id || !out) return false;
    for (size_t i = 0; i < board->fen_count; ++i)
        if (board->fences[i].fence_id && strcmp(board->fences[i].fence_id, id) == 0) {
            ca_geofence_copy(out, &board->fences[i]);
            return true;
        }
    return false;
}

static double sc_deg_to_rad(double d) { return d * SC_PI / 180.0; }

static double sc_haversine_m(double a_lat, double a_lon, double b_lat, double b_lon) {
    const double R = 6371000.0; /* 6_371_000 */
    double d_lat = sc_deg_to_rad(b_lat - a_lat);
    double d_lon = sc_deg_to_rad(b_lon - a_lon);
    double s1 = sin(d_lat / 2.0);
    double s2 = sin(d_lon / 2.0);
    double a = s1 * s1 + cos(sc_deg_to_rad(a_lat)) * cos(sc_deg_to_rad(b_lat)) * s2 * s2;
    double c = 2.0 * atan2(sqrt(a), sqrt(1.0 - a));
    return R * c;
}

bool ca_child_safety_board_is_inside_any_fence(ca_child_safety_board_t *board,
                                               double lat, double lon) {
    if (!board) return false;
    for (size_t i = 0; i < board->fen_count; ++i) {
        const ca_geofence_t *g = &board->fences[i];
        if (sc_haversine_m(g->centre_lat, g->centre_lon, lat, lon) <= g->radius_meters)
            return true;
    }
    return false;
}

bool ca_child_safety_board_record_check_in(ca_child_safety_board_t *board,
                                           const ca_check_in_t *c) {
    if (!board || !c) return false;
    if (board->chk_count == board->chk_cap) {
        size_t nc = board->chk_cap ? board->chk_cap * 2 : 8;
        void *n = realloc(board->checkins, nc * sizeof(*board->checkins));
        if (!n) return false;
        board->checkins = n; board->chk_cap = nc;
    }
    ca_check_in_copy(&board->checkins[board->chk_count], c);
    board->chk_count++;
    return true;
}

ca_check_in_t *ca_child_safety_board_recent_check_ins(ca_child_safety_board_t *board,
                                                      const char *child_id, int limit,
                                                      size_t *out_count) {
    if (out_count) *out_count = 0;
    /* C# throws ArgumentOutOfRangeException for limit<=0 (checked before the
     * childId filter). */
    if (limit <= 0) { if (out_count) *out_count = SIZE_MAX; return NULL; }
    if (!board || !child_id) { if (out_count) *out_count = SIZE_MAX; return NULL; }

    /* pick matching indices in source order */
    size_t *pick = (size_t *)malloc(board->chk_count ? board->chk_count * sizeof(size_t) : 1);
    if (!pick) { if (out_count) *out_count = SIZE_MAX; return NULL; }
    size_t m = 0;
    for (size_t i = 0; i < board->chk_count; ++i)
        if (board->checkins[i].child_id && strcmp(board->checkins[i].child_id, child_id) == 0)
            pick[m++] = i;
    if (m == 0) { free(pick); return NULL; }

    /* OrderByDescending(c => c.AtUtc) — stable descending. */
    for (size_t i = 1; i < m; ++i) {
        size_t cur = pick[i];
        int64_t key = board->checkins[cur].at_utc_ms;
        size_t j = i;
        while (j > 0 && board->checkins[pick[j - 1]].at_utc_ms < key) { pick[j] = pick[j - 1]; --j; }
        pick[j] = cur;
    }

    size_t take = m < (size_t)limit ? m : (size_t)limit;
    ca_check_in_t *res = (ca_check_in_t *)calloc(take, sizeof(*res));
    if (!res) { free(pick); if (out_count) *out_count = SIZE_MAX; return NULL; }
    for (size_t i = 0; i < take; ++i) ca_check_in_copy(&res[i], &board->checkins[pick[i]]);
    free(pick);
    if (out_count) *out_count = take;
    return res;
}

File: go/safety_child.go
// safety_child.go
//
// Ports the CircleAI.Safety.Child module (namespace CircleAI.SafetyChild):
//   Records:    TrustedAdult, Geofence, CheckIn (ChildSafetyPrimitives.cs)
//   Interfaces: IChildSafetyBoard (ChildSafetyPrimitives.cs)
//   Impls:      InMemoryChildSafetyBoard (ChildSafetyPrimitives.cs)
//   Constants:  SafetyChildDomainContext (SafetyChildDomainContext.cs)
//
// The Child-Safety vertical is safeguarding: a priority-ordered trusted-adult
// ring, named geofences with a Haversine containment test, and per-child
// check-in history, held in a thread-safe in-memory board.
//
// Records are value structs in Go and are therefore never nil; the C#
// ArgumentNullException.ThrowIfNull guards on the mutators are structurally
// unrepresentable and so are dropped. RecentCheckIns preserves the C#
// ArgumentOutOfRangeException on a non-positive limit by panicking (a contract
// violation, matching the thrown exception). The Haversine implementation is a
// faithful translation of the C# private HaversineMeters (R = 6_371_000 m).

package circleai

import (
	"fmt"
	"math"
	"sort"
	"sync"
	"time"
)

// TrustedAdult is a member of a child's trusted-adult ring. Ports TrustedAdult.
type TrustedAdult struct {
	// AdultID is the stable identifier.
	AdultID string
	// Name is the adult's name.
	Name string
	// Phone is the adult's phone number.
	Phone string
	// Relationship describes the relationship to the child.
	Relationship string
	// RingPriority orders the ring ascending (lower == contacted first).
	RingPriority int
}

// Geofence is a named circular geofence. Ports Geofence.
type Geofence struct {
	// FenceID is the stable identifier.
	FenceID string
	// Name is the fence's name.
	Name string
	// CentreLat is the fence centre latitude in degrees.
	CentreLat float64
	// CentreLon is the fence centre longitude in degrees.
	CentreLon float64
	// RadiusMeters is the fence radius in metres.
	RadiusMeters float64
}

// CheckIn is a child check-in event. Ports CheckIn. Lat / Lon are pointers to
// model the C# nullable double? (nil == no coordinate).
type CheckIn struct {
	// ChildID is the child the check-in belongs to.
	ChildID string
	// Status is a free-form status label.
	Status string
	// Lat is the optional check-in latitude.
	Lat *float64
	// Lon is the optional check-in longitude.
	Lon *float64
	// AtUTC is when the check-in occurred.
	AtUTC time.Time
}

// IChildSafetyBoard is the child-safeguarding board contract. Ports
// IChildSafetyBoard.
type IChildSafetyBoard interface {
	// AddAdult adds (or replaces, keyed by AdultID) a trusted adult.
	AddAdult(a TrustedAdult)
	// RingOrdered returns the trusted-adult ring ordered by ascending priority.
	RingOrdered() []TrustedAdult
	// DefineGeofence adds (or replaces, keyed by FenceID) a geofence.
	DefineGeofence(g Geofence)
	// GetGeofence returns the geofence with id, or nil when unknown.
	GetGeofence(id string) *Geofence
	// IsInsideAnyFence reports whether (lat, lon) falls within any geofence.
	IsInsideAnyFence(lat, lon float64) bool
	// RecordCheckIn records a check-in event.
	RecordCheckIn(c CheckIn)
	// RecentCheckIns returns up to limit of childID's check-ins, newest-first.
	RecentCheckIns(childID string, limit int) []CheckIn
}

// InMemoryChildSafetyBoard is a thread-safe in-memory IChildSafetyBoard. Ports
// InMemoryChildSafetyBoard.
type InMemoryChildSafetyBoard struct {
	mu       sync.Mutex
	adults   map[string]TrustedAdult
	fences   map[string]Geofence
	checkIns []CheckIn
}

// NewInMemoryChildSafetyBoard constructs an empty board.
func NewInMemoryChildSafetyBoard() *InMemoryChildSafetyBoard {
	return &InMemoryChildSafetyBoard{
		adults: make(map[string]TrustedAdult),
		fences: make(map[string]Geofence),
	}
}

// AddAdult records or replaces a trusted adult keyed by AdultID. Ports
// InMemoryChildSafetyBoard.AddAdult.
func (b *InMemoryChildSafetyBoard) AddAdult(a TrustedAdult) {
	b.mu.Lock()
	b.adults[a.AdultID] = a
	b.mu.Unlock()
}

// RingOrdered returns the ring ordered by ascending RingPriority. Ports
// InMemoryChildSafetyBoard.RingOrdered (OrderBy(a => a.RingPriority)).
func (b *InMemoryChildSafetyBoard) RingOrdered() []TrustedAdult {
	b.mu.Lock()
	defer b.mu.Unlock()
	out := make([]TrustedAdult, 0, len(b.adults))
	for _, a := range b.adults {
		out = append(out, a)
	}
	sort.SliceStable(out, func(i, j int) bool { return out[i].RingPriority < out[j].RingPriority })
	return out
}

// DefineGeofence records or replaces a geofence keyed by FenceID. Ports
// InMemoryChildSafetyBoard.DefineGeofence.
func (b *InMemoryChildSafetyBoard) DefineGeofence(g Geofence) {
	b.mu.Lock()
	b.fences[g.FenceID] = g
	b.mu.Unlock()
}

// GetGeofence returns the geofence with id, or nil when unknown. Ports
// InMemoryChildSafetyBoard.GetGeofence.
func (b *InMemoryChildSafetyBoard) GetGeofence(id string) *Geofence {
	b.mu.Lock()
	defer b.mu.Unlock()
	g, ok := b.fences[id]
	if !ok {
		return nil
	}
	return &g
}

// IsInsideAnyFence reports whether (lat, lon) is within any defined geofence,
// by Haversine distance. Ports InMemoryChildSafetyBoard.IsInsideAnyFence.
func (b *InMemoryChildSafetyBoard) IsInsideAnyFence(lat, lon float64) bool {
	b.mu.Lock()
	defer b.mu.Unlock()
	for _, g := range b.fences {
		if haversineMeters(g.CentreLat, g.CentreLon, lat, lon) <= g.RadiusMeters {
			return true
		}
	}
	return false
}

// RecordCheckIn appends a check-in event. Ports
// InMemoryChildSafetyBoard.RecordCheckIn.
func (b *InMemoryChildSafetyBoard) RecordCheckIn(c CheckIn) {
	b.mu.Lock()
	b.checkIns = append(b.checkIns, c)
	b.mu.Unlock()
}

// RecentCheckIns returns up to limit of childID's check-ins, newest-first. It
// panics when limit <= 0, mirroring the C# ArgumentOutOfRangeException. Ports
// InMemoryChildSafetyBoard.RecentCheckIns.
func (b *InMemoryChildSafetyBoard) RecentCheckIns(childID string, limit int) []CheckIn {
	if limit <= 0 {
		panic(fmt.Sprintf("limit out of range: %d (must be positive)", limit))
	}
	b.mu.Lock()
	defer b.mu.Unlock()
	matched := make([]CheckIn, 0)
	for _, c := range b.checkIns {
		if c.ChildID == childID {
			matched = append(matched, c)
		}
	}
	sort.SliceStable(matched, func(i, j int) bool { return matched[i].AtUTC.After(matched[j].AtUTC) })
	if len(matched) > limit {
		matched = matched[:limit]
	}
	return matched
}

// haversineMeters returns the great-circle distance in metres between two
// lat/lon points. Faithful translation of the C# private HaversineMeters
// (R = 6_371_000 m; degrees→radians; atan2 of the half-angle sines).
func haversineMeters(aLat, aLon, bLat, bLon float64) float64 {
	const r = 6_371_000.0
	degToRad := func(d float64) float64 { return d * math.Pi / 180.0 }
	dLat := degToRad(bLat - aLat)
	dLon := degToRad(bLon - aLon)
	s1 := math.Sin(dLat / 2)
	s2 := math.Sin(dLon / 2)
	a := s1*s1 + math.Cos(degToRad(aLat))*math.Cos(degToRad(bLat))*s2*s2
	c := 2 * math.Atan2(math.Sqrt(a), math.Sqrt(1-a))
	return r * c
}

// ─── SafetyChildDomainContext ──────────────────────────────────────────────

// safetyChildDomainContext holds the static domain descriptor for the
// Child-Safety vertical. Ports the static class SafetyChildDomainContext.
type safetyChildDomainContext struct{}

// SafetyChildDomainContext is the singleton domain descriptor accessor. Ports
// SafetyChildDomainContext.
var SafetyChildDomainContext = safetyChildDomainContext{}

// SystemPromptSnippet returns the domain system-prompt preamble. Ports
// SafetyChildDomainContext.SystemPromptSnippet.
func (safetyChildDomainContext) SystemPromptSnippet() string {
	return "[DOMAIN: Safety.Child] Child safety and safeguarding assistant for parents and educators. Help with online safety education, age-appropriate device rules, recognising grooming signs, reporting abuse, and digital literacy. Always prioritise child welfare. IMPORTANT: For immediate child safety concerns, contact SAPS (10111) or Childline (116). Compliance: Children's Act 38/2005, POPIA (children's data), FILMS_PUBLICATIONS_ACT, Cybercrimes Act."
}

// ComplianceFlags returns the compliance flags for the Child-Safety vertical.
// Ports SafetyChildDomainContext.ComplianceFlags.
func (safetyChildDomainContext) ComplianceFlags() []string {
	return []string{"Childrens_Act_38_2005", "POPIA_Children", "Films_Publications_Act", "Cybercrimes_Act", "Emergency_116"}
}

// SuggestedTools returns the suggested tool ids for the Child-Safety vertical.
// Ports SafetyChildDomainContext.SuggestedTools.
func (safetyChildDomainContext) SuggestedTools() []string {
	return []string{"parental_controls", "web_search", "document_editor", "reporting_tools"}
}

// Compile-time assertion that the implementation satisfies the contract.
var _ IChildSafetyBoard = (*InMemoryChildSafetyBoard)(nil)
